third_party/terraform/utils/batcher.go

```diff
@@ -0,0 +1,204 @@
+
+// Function type for sending a batch request
+type batcherSendFunc func(resourceName string, body interface{}) (interface{}, error)
+
+// batchResponse bundles an API response (data, error) tuple.
+type batchResponse struct {
+	body interface{}
+	err  error
+}
+
+// startedBatch refers to a processed batch whose timer to send the request has
+// already been started. The responses for the request is sent to each listener
+// channel, representing parallel callers that are waiting on requests
+// combined into this batch.
+type startedBatch struct {
+	*BatchRequest
+
+	listeners []chan batchResponse
+	timer     *time.Timer
+}
+
+// batchingConfig contains user configuration for controlling batch requests.
+type batchingConfig struct {
+	sendAfter       time.Duration
+	disableBatching bool
+}
+
+// Initializes a new batcher
+func NewRequestBatcher(config *batchingConfig) *RequestBatcher {
+	return &RequestBatcher{
+		batchingConfig: config,
+		batches:        make(map[string]*startedBatch),
+	}
+}
+
+// SendRequestWithTimeout is expected to be called per parallel call.
+// It manages waiting on the result of a batch request.
+//
+// Batch requests are grouped by the given batchKey, and thus batchKey
+// should be unique to the type of batch request being sent and the affected
+// third-party resource.
+func (b *RequestBatcher) SendRequestWithTimeout(batchKey string, request *BatchRequest, timeout time.Duration) (interface{}, error) {
+	if request == nil {
+		return nil, fmt.Errorf("error, cannot request batching for nil BatchRequest")
+	}
+	if request.CombineF == nil {
+		return nil, fmt.Errorf("error, cannot request batching for BatchRequest with nil CombineF")
+	}
+	if request.SendF == nil {
+		return nil, fmt.Errorf("error, cannot request batching for BatchRequest with nil SendF")
+	}
+	if b.disableBatching {
+		log.Printf("[DEBUG] Batching is disabled, sending single request for %q", request.DebugId)
+		return request.SendF(batchKey, request.Body)
+	}
+
+	respCh, err := b.startBatchRequest(batchKey, request)
+	if err != nil {
+		return nil, fmt.Errorf("error adding request to batch: %s", err)
+	}
+
+	timer := time.NewTimer(timeout)
+
+	select {
+	case resp := <-respCh:
+		if resp.err != nil {
+			return nil, fmt.Errorf("Batch %q for request %q returned error: %v", batchKey, request.DebugId, resp.err)
+		}
+		return resp.body, nil
+	case <-timer.C:
+		break
+	}
+	return nil, fmt.Errorf("Request %s timed out after %v", batchKey, timeout)
+}
+
+// startBatchRequest manages batching logic that access shared information
+// (i.e. existing batches)
+func (b *RequestBatcher) startBatchRequest(batchKey string, newRequest *BatchRequest) (<-chan batchResponse, error) {
+	b.Lock()
+	defer b.Unlock()
+
+	// The calling goroutine will need a channel to wait on for a response.
+	respCh := make(chan batchResponse, 1)
+
+	// If batch already exists, combine this request into existing request.
+	if batch, ok := b.batches[batchKey]; ok {
```

Comment:
minor: I might wrap this if block in its own method. That way it could read more like a guard clause rather than a block with early returns of both error & success.
Something like `return mergeExistingBatch(stuff) if batch exists`